File: FunctionalArray.hs
import GHC.Generics (Generic)
import Control.DeepSeq (NFData, rnf, deepseq)
import Criterion.Main
import System.Random (randomRIO)
import GHC.Stats (getRTSStats, RTSStats(allocated_bytes), getRTSStatsEnabled)
import System.Mem (performGC)
import System.Clock (Clock(Monotonic), getTime, diffTimeSpec, toNanoSecs)

data FunctionalArray a = FunctionalArray [a] deriving (Show, Generic)


instance NFData a => NFData (FunctionalArray a) where
    rnf (FunctionalArray elements) = rnf elements


get :: Int -> FunctionalArray a -> Maybe a
get index (FunctionalArray elements)
  | index >= 0 && index < length elements = Just (elements !! index)
  | otherwise = Nothing


update :: Int -> a -> FunctionalArray a -> FunctionalArray a
update index value (FunctionalArray elements)
  | index >= 0 && index < length elements =
      let (before, _:after) = splitAt index elements
      in FunctionalArray (before ++ [value] ++ after)
  | otherwise = error "Index out of bounds"


append :: a -> FunctionalArray a -> FunctionalArray a
append value (FunctionalArray elements) = FunctionalArray (elements ++ [value])


prepend :: a -> FunctionalArray a -> FunctionalArray a
prepend value (FunctionalArray elements) = FunctionalArray (value : elements)


lengthArray :: FunctionalArray a -> Int
lengthArray (FunctionalArray elements) = length elements


slice :: Int -> Int -> FunctionalArray a -> FunctionalArray a
slice start end (FunctionalArray elements) = FunctionalArray (take (end - start) (drop start elements))


mapArray :: (a -> b) -> FunctionalArray a -> FunctionalArray b
mapArray f (FunctionalArray elements) = FunctionalArray (map f elements)


filterArray :: (a -> Bool) -> FunctionalArray a -> FunctionalArray a
filterArray p (FunctionalArray elements) = FunctionalArray (filter p elements)


foldArray :: (b -> a -> b) -> b -> FunctionalArray a -> b
foldArray f acc (FunctionalArray elements) = foldl f acc elements


insertFirst :: a -> FunctionalArray a -> FunctionalArray a
insertFirst = prepend


insertLast :: a -> FunctionalArray a -> FunctionalArray a
insertLast = append


deleteFirst :: FunctionalArray a -> FunctionalArray a
deleteFirst (FunctionalArray []) = FunctionalArray []
deleteFirst (FunctionalArray (_:xs)) = FunctionalArray xs


deleteLast :: FunctionalArray a -> FunctionalArray a
deleteLast (FunctionalArray []) = FunctionalArray []
deleteLast (FunctionalArray xs) = FunctionalArray (init xs)


reverse1 :: FunctionalArray a -> FunctionalArray a
reverse1 (FunctionalArray elements) = FunctionalArray (reverse elements)


merge :: FunctionalArray a -> FunctionalArray a -> FunctionalArray a
merge (FunctionalArray xs) (FunctionalArray ys) = FunctionalArray (xs ++ ys)


findByIndex :: Int -> FunctionalArray a -> Maybe a
findByIndex = get


fromList :: [a] -> FunctionalArray a
fromList = FunctionalArray


generateRandomList :: Int -> Int -> Int -> IO [Int]
generateRandomList n low high = sequence (replicate n (randomRIO (low, high)))


measureTimeAndMemory :: NFData a => IO a -> IO ()
measureTimeAndMemory action = do
    statsEnabled <- getRTSStatsEnabled
    if not statsEnabled
        then putStrLn "RTS Stats are not enabled. Run GHCi with +RTS -T."
        else do
            performGC
            
            -- Start time and memory
            startStats <- getRTSStats
            let startMem = allocated_bytes startStats
            startTime <- getTime Monotonic
            
            -- Perform the action
            result <- action
            result `deepseq` return ()
            
            -- End time and memory
            endStats <- getRTSStats
            let endMem = allocated_bytes endStats
            endTime <- getTime Monotonic
            
            -- Calculate time and memory usage
            let elapsedTime = fromIntegral (toNanoSecs (diffTimeSpec startTime endTime)) / (10^6) -- Convert to milliseconds
            let memoryUsed = endMem - startMem
            
            -- Print results
            putStrLn $ "Execution time: " ++ show elapsedTime ++ " ms"
            putStrLn $ "Memory used: " ++ show memoryUsed ++ " bytes"




main :: IO ()
main = runMemoryProfiling

runMemoryProfiling :: IO ()
runMemoryProfiling = do
    randomList <- generateRandomList 10000 1 10000
    let functionalArray = fromList randomList
    let functionalArray2 = fromList randomList

    putStrLn "Performance for operations:"
    measureTimeAndMemory $ return $ foldArray (\acc x -> append x acc) (FunctionalArray []) randomList
    putStrLn "Append profiling completed."

    measureTimeAndMemory $ return $ foldArray (\acc x -> insertFirst x acc) (FunctionalArray []) randomList
    putStrLn "InsertFirst profiling completed."

    measureTimeAndMemory $ return $ foldArray (\acc x -> insertLast x acc) (FunctionalArray []) randomList
    putStrLn "InsertLast profiling completed."

    measureTimeAndMemory $ return $ foldArray (\acc _ -> deleteFirst acc) functionalArray [1..10000]
    putStrLn "DeleteFirst profiling completed."

    measureTimeAndMemory $ return $ foldArray (\acc _ -> deleteLast acc) functionalArray [1..10000]
    putStrLn "DeleteLast profiling completed."

    measureTimeAndMemory $ return $ reverse1 functionalArray
    putStrLn "Reverse profiling completed."

    measureTimeAndMemory $ return $ merge functionalArray functionalArray2
    putStrLn "Merge profiling completed."

    measureTimeAndMemory $ return $ findByIndex 5000 functionalArray
    putStrLn "FindByIndex profiling completed."

    measureTimeAndMemory $ return $ get 5000 functionalArray
    putStrLn "Get profiling completed."

    measureTimeAndMemory $ return $ update 5000 9999 functionalArray
    putStrLn "Update profiling completed."

    measureTimeAndMemory $ return $ prepend 0 functionalArray
    putStrLn "Prepend profiling completed."

    measureTimeAndMemory $ return $ lengthArray functionalArray
    putStrLn "Length profiling completed."

    measureTimeAndMemory $ return $ slice 1000 2000 functionalArray
    putStrLn "Slice profiling completed."

    measureTimeAndMemory $ return $ mapArray (+1) functionalArray
    putStrLn "MapArray profiling completed."

    measureTimeAndMemory $ return $ filterArray even functionalArray
    putStrLn "FilterArray profiling completed."

    measureTimeAndMemory $ return $ foldArray (+) 0 functionalArray
    putStrLn "FoldArray profiling completed."

generateRandomList :: Int -> Int -> Int -> IO [Int]
generateRandomList n low high = sequence (replicate n (randomRIO (low, high)))

measureTimeAndMemory :: NFData a => IO a -> IO ()
measureTimeAndMemory action = do
    statsEnabled <- getRTSStatsEnabled
    if not statsEnabled
        then putStrLn "RTS Stats are not enabled. Run GHCi with +RTS -T."
        else do
            performGC -- Force garbage collection for accurate measurement
            
            -- Start time and memory
            startStats <- getRTSStats
            let startMem = allocated_bytes startStats
            startTime <- getTime Monotonic
            
            -- Perform the action
            result <- action
            result `deepseq` return ()
            
            -- End time and memory
            endStats <- getRTSStats
            let endMem = allocated_bytes endStats
            endTime <- getTime Monotonic
            
            -- Calculate time and memory usage
            let elapsedTime = fromIntegral (toNanoSecs (diffTimeSpec startTime endTime)) / (10^6) -- Convert to milliseconds
            let memoryUsed = endMem - startMem
            
            -- Print results
            putStrLn $ "Execution time: " ++ show elapsedTime ++ " ms"
            putStrLn $ "Memory used: " ++ show memoryUsed ++ " bytes"

File: FunctionalArray.py
import random
import time
import tracemalloc
from typing import List, Callable, Optional, TypeVar, Generic

T = TypeVar('T')
U = TypeVar('U')

class FunctionalArray(Generic[T]):
    def __init__(self, elements: List[T]):
        self.elements = elements

    def __repr__(self):
        return f"FunctionalArray({self.elements})"

    def get(self, index: int) -> Optional[T]:
        if 0 <= index < len(self.elements):
            return self.elements[index]
        return None

    def update(self, index: int, value: T) -> 'FunctionalArray[T]':
        if 0 <= index < len(self.elements):
            new_elements = self.elements[:index] + [value] + self.elements[index + 1:]
            return FunctionalArray(new_elements)
        raise IndexError("Index out of bounds")

    def append(self, value: T) -> 'FunctionalArray[T]':
        return FunctionalArray(self.elements + [value])

    def prepend(self, value: T) -> 'FunctionalArray[T]':
        return FunctionalArray([value] + self.elements)

    def length(self) -> int:
        return len(self.elements)

    def slice(self, start: int, end: int) -> 'FunctionalArray[T]':
        return FunctionalArray(self.elements[start:end])

    def map_array(self, func: Callable[[T], U]) -> 'FunctionalArray[U]':
        return FunctionalArray(list(map(func, self.elements)))

    def filter_array(self, predicate: Callable[[T], bool]) -> 'FunctionalArray[T]':
        return FunctionalArray(list(filter(predicate, self.elements)))

    def fold_array(self, func: Callable[[U, T], U], acc: U) -> U:
        result = acc
        for element in self.elements:
            result = func(result, element)
        return result

    def insert_first(self, value: T) -> 'FunctionalArray[T]':
        return self.prepend(value)

    def insert_last(self, value: T) -> 'FunctionalArray[T]':
        return self.append(value)

    def delete_first(self) -> 'FunctionalArray[T]':
        return FunctionalArray(self.elements[1:])

    def delete_last(self) -> 'FunctionalArray[T]':
        return FunctionalArray(self.elements[:-1])

    def reverse(self) -> 'FunctionalArray[T]':
        return FunctionalArray(self.elements[::-1])

    def merge(self, other: 'FunctionalArray[T]') -> 'FunctionalArray[T]':
        return FunctionalArray(self.elements + other.elements)

    def find_by_index(self, index: int) -> Optional[T]:
        return self.get(index)

    @staticmethod
    def from_list(elements: List[T]) -> 'FunctionalArray[T]':
        return FunctionalArray(elements)

def generate_random_list(n: int, low: int, high: int) -> List[int]:
    return [random.randint(low, high) for _ in range(n)]

def measure_time_and_memory(func: Callable[[], T]) -> None:
    tracemalloc.start()
    start_time = time.time()
    start_memory = tracemalloc.get_traced_memory()[1]

    result = func()

    end_time = time.time()
    end_memory = tracemalloc.get_traced_memory()[1]
    tracemalloc.stop()

    elapsed_time = (end_time - start_time) * 1000  # Convert to milliseconds
    memory_used = end_memory - start_memory

    print(f"Execution time: {elapsed_time:.2f} ms")
    print(f"Memory used: {memory_used} bytes")

def run_memory_profiling() -> None:
    random_list = generate_random_list(10000, 1, 10000)
    functional_array = FunctionalArray.from_list(random_list)
    functional_array2 = FunctionalArray.from_list(random_list)

    print("Performance for operations:")
    measure_time_and_memory(lambda: FunctionalArray.from_list(random_list).fold_array(lambda acc, x: acc.append(x), FunctionalArray([])))
    print("Append profiling completed.")

    measure_time_and_memory(lambda: FunctionalArray.from_list(random_list).fold_array(lambda acc, x: acc.insert_first(x), FunctionalArray([])))
    print("InsertFirst profiling completed.")

    measure_time_and_memory(lambda: FunctionalArray.from_list(random_list).fold_array(lambda acc, x: acc.insert_last(x), FunctionalArray([])))
    print("InsertLast profiling completed.")

    measure_time_and_memory(lambda: functional_array.fold_array(lambda acc, _: acc.delete_first(), functional_array))
    print("DeleteFirst profiling completed.")

    measure_time_and_memory(lambda: functional_array.fold_array(lambda acc, _: acc.delete_last(), functional_array))
    print("DeleteLast profiling completed.")

    measure_time_and_memory(lambda: functional_array.reverse())
    print("Reverse profiling completed.")

    measure_time_and_memory(lambda: functional_array.merge(functional_array2))
    print("Merge profiling completed.")

    measure_time_and_memory(lambda: functional_array.find_by_index(5000))
    print("FindByIndex profiling completed.")

    measure_time_and_memory(lambda: functional_array.map_array(lambda x: x * 2))
    print("Map profiling completed.")

    measure_time_and_memory(lambda: functional_array.filter_array(lambda x: x % 2 == 0))
    print("Filter profiling completed.")

    measure_time_and_memory(lambda: functional_array.slice(5000, 6000))
    print("Slice profiling completed.")

    measure_time_and_memory(lambda: functional_array.length())
    print("Length profiling completed.")

    measure_time_and_memory(lambda: functional_array.get(5000))
    print("Get profiling completed.")

    measure_time_and_memory(lambda: functional_array.update(5000, 10000))
    print("Update profiling completed.")

    measure_time_and_memory(lambda: functional_array.fold_array(10000))
    print("Fold profiling completed.")

    measure_time_and_memory(lambda: FunctionalArray.prepend(random_list))
    print("Prepend profiling completed.")
    


if __name__ == "__main__":
    run_memory_profiling()